internal/controller/pkg/revision/deployment.go

```diff
@@ -265,136 +210,250 @@ func buildProviderDeployment(provider *pkgmetav1.Provider, revision v1.PackageRe
+		templateLabels[k] = v
+	}
+	d.Spec.Template.Labels = templateLabels
+
+	svc := getService(revision.GetName(), namespace, []metav1.OwnerReference{meta.AsController(meta.TypedReferenceTo(revision, v1.ProviderRevisionGroupVersionKind))}, d.Spec.Selector.MatchLabels)
+
+	return s, d, svc, secSer, secCli
+}
+
+func buildFunctionDeployment(function *pkgmetav1alpha1.Function, revision v1.PackageRevision, cc *v1alpha1.ControllerConfig, namespace string, pullSecrets []corev1.LocalObjectReference) (*corev1.ServiceAccount, *appsv1.Deployment, *corev1.Service, *corev1.Secret) {
+	s := &corev1.ServiceAccount{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            revision.GetName(),
+			Namespace:       namespace,
+			OwnerReferences: []metav1.OwnerReference{meta.AsController(meta.TypedReferenceTo(revision, v1alpha1.FunctionRevisionGroupVersionKind))},
+		},
+		ImagePullSecrets: pullSecrets,
+	}
+	sec := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:            *revision.GetTLSServerSecretName(),
+			Namespace:       namespace,
+			OwnerReferences: revision.GetOwnerReferences(),
+		},
+	}
+
+	pullPolicy := corev1.PullIfNotPresent
+	if revision.GetPackagePullPolicy() != nil {
+		pullPolicy = *revision.GetPackagePullPolicy()
+	}
+
+	image := revision.GetSource()
+	if function.Spec.Image != nil {
+		image = *function.Spec.Image
+	}
+
+	d := &appsv1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      revision.GetName(),
+			Namespace: namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				meta.AsController(meta.TypedReferenceTo(revision, v1alpha1.FunctionRevisionGroupVersionKind)),
+			},
+		},
+		Spec: appsv1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"pkg.crossplane.io/revision": revision.GetName(),
+					"pkg.crossplane.io/function": function.GetName(),
+				},
+			},
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      function.GetName(),
+					Namespace: namespace,
+				},
+				Spec: corev1.PodSpec{
+					SecurityContext: &corev1.PodSecurityContext{
+						RunAsNonRoot: &runAsNonRoot,
+						RunAsUser:    &runAsUser,
+						RunAsGroup:   &runAsGroup,
+					},
+					ServiceAccountName: s.GetName(),
+					ImagePullSecrets:   revision.GetPackagePullSecrets(),
+					Containers: []corev1.Container{
+						{
+							Name:            function.GetName(),
+							Image:           image,
+							ImagePullPolicy: pullPolicy,
+							SecurityContext: &corev1.SecurityContext{
+								RunAsUser:                &runAsUser,
+								RunAsGroup:               &runAsGroup,
+								AllowPrivilegeEscalation: &allowPrivilegeEscalation,
+								Privileged:               &privileged,
+								RunAsNonRoot:             &runAsNonRoot,
+							},
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          promPortName,
+									ContainerPort: promPortNumber,
+								},
```

Comment:
"metrics"? prometheus? 🤔

Reply:
yes, it's for metrics

Reply:
but that's the only port being defined, we are missing the 9443, svc is still working fine, but it's usually better to define ports on containers nonetheless (dropping a message here too just not to forget)